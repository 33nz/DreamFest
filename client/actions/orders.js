export const FETCH_ORDERS_PENDING = 'FETCH_ORDERS_PENDING'
export const FETCH_ORDERS_SUCCESS = 'FETCH_ORDERS_SUCCESS'
export const PATCH_ORDER_PENDING = 'PATCH_ORDER_PENDING'
export const PATCH_ORDER_SUCCESS = 'PATCH_ORDER_SUCCESS'

export function fetchOrdersPending () {
  return {
    type: FETCH_ORDERS_PENDING
  }
}

export function fetchOrdersSuccess (orders) {
  return {
    type: FETCH_ORDERS_SUCCESS,
    orders: orders
  }
}

export function patchOrderPending () {
  return {
    type: PATCH_ORDER_PENDING
  }
}

export function patchOrderSuccess (order) {
  return {
    type: PATCH_ORDER_SUCCESS,
    order: order
  }
}
