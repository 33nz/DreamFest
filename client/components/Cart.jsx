import React from 'react'
import { connect } from 'react-redux'

import CartItem from './CartItem'

import {
  deleteFromCart,
  updateCart
} from '../actions/cart'

class Cart extends React.Component {
  state = {
    cart: this.props.cart
  }

  update = (id, quantity) => {
    this.setState({
      cart: this.state.cart.map(item => {
        if (item.id === id) item.quantity = Number(quantity)
        return item
      })
    })
  }

  deleteItem = (id) => {
    const cart = this.state.cart.filter(item => item.id !== id)
    this.setState({ cart })
    this.props.deleteFromCart(id)
  }

  keepShopping = () => {
    this.props.history.push('/')
  }

  render () {
    return (
      <div className='cart'>
        <table>
          <thead>
            <tr>
              <td>Product</td>
              <td>Quantity</td>
              <td>Remove</td>
            </tr>
          </thead>
          <tbody>
            {this.props.cart.map((item, id) => {
              return (
                <CartItem key={id}
                          item={item}
                          update={this.update}
                          deleteFromCart={this.deleteItem}
                />)
            })}
          </tbody>
        </table>
        <p className='actions'>
          <button onClick={this.keepShopping}>Continue shopping</button>
          <button onClick={() => this.props.updateCart(this.state.cart)}>Update</button>
          <button className='button-primary'>Place Order</button>
        </p>
      </div>
    )
  }
}

const mapStateToProps = (state) => {
  return {
    cart: state.cart
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    deleteFromCart: (id) => dispatch(deleteFromCart(id)),
    updateCart: (cart) => dispatch(updateCart(cart))
  }
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Cart)
