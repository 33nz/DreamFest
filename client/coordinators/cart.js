export function updateCartItem (updateInfo, updateCartDispatcher) {
  const isValid = !isNaN(Number(updateInfo.quantity))
  isValid && updateCartDispatcher(updateInfo)
}
