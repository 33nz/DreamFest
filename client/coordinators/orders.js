import requestor from '../consume'

export function getOrders (dispatchers, consume = requestor) {
  const { fetchOrdersPending, fetchOrdersSuccess, showError } = dispatchers
  fetchOrdersPending()
  return consume('/orders')
    .then(res => {
      const orders = res.body
      fetchOrdersSuccess(orders)
      return null
    })
    .catch(err => {
      showError(err.message)
    })
}
