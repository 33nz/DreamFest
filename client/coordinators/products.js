import requestor from '../consume'

export function getProducts (dispatchers, consume = requestor) {
  const { fetchProductsPending, fetchProductsSuccess, showError } = dispatchers
  fetchProductsPending()
  return consume('/products')
    .then(res => {
      const products = res.body
      fetchProductsSuccess(products)
      return null
    })
    .catch(err => {
      showError(err.message)
    })
}
