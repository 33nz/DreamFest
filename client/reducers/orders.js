import {
  FETCH_ORDERS_SUCCESS,
  POST_ORDER_SUCCESS,
  PATCH_ORDER_SUCCESS,
  DELETE_ORDER_SUCCESS
} from '../actions/orders'

const orders = (state = [], action) => {
  switch (action.type) {
    case FETCH_ORDERS_SUCCESS:
      return action.orders

    case POST_ORDER_SUCCESS:
      return [ ...state, action.order ]

    case PATCH_ORDER_SUCCESS:
      return state.map(ord => (ord.id === action.order.id) ? action.order : ord)

    case DELETE_ORDER_SUCCESS:
      return state.filter(order => order.id !== action.orderId)

    default:
      return state
  }
}

export default orders
