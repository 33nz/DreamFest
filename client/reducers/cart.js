import {
  ADD_TO_CART,
  DELETE_FROM_CART,
  UPDATE_CART } from '../actions/cart'
import { POST_ORDER_SUCCESS } from '../actions/orders'
import { getNewCart } from './reducer-helpers'

const cart = (state = [], action) => {
  switch (action.type) {
    case ADD_TO_CART:
      return getNewCart(state, action.product)

    case DELETE_FROM_CART:
      return state.filter(item => item.id !== action.id)

    case UPDATE_CART:
      return action.cart

    case POST_ORDER_SUCCESS:
      return []

    default:
      return state
  }
}

export default cart
