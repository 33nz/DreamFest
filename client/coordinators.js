import { showError } from './actions/error'
import { requestProducts, receiveProducts } from './actions/products'

import { fetchProducts } from './api'

export function getProducts (dispatch) {
  dispatch(requestProducts())
  fetchProducts()
    .then(products => {
      dispatch(receiveProducts(products))
    })
    .catch(err => {
      dispatch(showError(err.message))
    })
}
