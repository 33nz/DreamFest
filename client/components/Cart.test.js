import React from 'react'
import { fireEvent, screen } from '@testing-library/react'
import { renderWithRouter } from '../testing/utils'
import '@testing-library/jest-dom'

import { Cart } from './Cart'

import mockCart from '../testing/mockCart'
import mockOrders from '../testing/mockOrders'

import { createOrder } from '../coordinators/cart'
import { placeOrder } from '../coordinators/orders'

jest.mock('../coordinators/cart', () => {
  return {
    createOrder: jest.fn()
  }
})

jest.mock('../coordinators/orders', () => {
  return {
    placeOrder: jest.fn()
  }
})

test('renders cart when in props', async () => {
  const mockChild = <>mock child fragment</>
  const { asFragment } = renderWithRouter(<Cart cart={mockCart}>
    {mockChild}</Cart>)

  const cartHeaders = await screen.getAllByRole('columnheader')
  const productCells = await screen.getAllByRole('cell')
  const quantityInputs = await screen.getAllByRole('textbox')

  expect(cartHeaders).toHaveLength(3)
  expect(productCells[0]).toHaveTextContent('mocked cart item 1')
  expect(quantityInputs).toHaveLength(3)
  expect(asFragment()).toMatchSnapshot()
})

test('calls placeOrder on button click', async () => {
  const order = mockOrders[0]
  const mockHistory = []
  createOrder.mockImplementation(() => order)

  renderWithRouter(<Cart
    cart={mockCart}
    history={mockHistory}
  />)
  const orderButton = await screen.getByRole('button', { name: 'Place Order' })
  fireEvent.click(orderButton)

  expect(createOrder).toHaveBeenCalledWith(mockCart)
  expect(placeOrder).toHaveBeenCalled()
  expect(placeOrder.mock.calls[0][0]).toBe(order)
})

test('renders a start shopping message when cart is empty', async () => {
  const { asFragment } = renderWithRouter(<Cart cart={[]}/>)
  const table = await screen.queryByRole('table')
  const emptyMessage = await screen.getByText('cart is empty', { exact: false })

  expect(table).not.toBeInTheDocument()
  expect(emptyMessage).toBeInTheDocument()
  expect(asFragment()).toMatchSnapshot()
})
