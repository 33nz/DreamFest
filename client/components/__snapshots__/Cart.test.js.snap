// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a start shopping message when cart is empty 1`] = `
<DocumentFragment>
  <p>
    Your cart is empty! Start shopping 
    <a
      href="/"
    >
      here
    </a>
  </p>
</DocumentFragment>
`;

exports[`renders cart when in props 1`] = `
<DocumentFragment>
  <div
    class="cart"
  >
    <table>
      <thead>
        <tr>
          <td
            role="columnheader"
          >
            Product
          </td>
          <td
            role="columnheader"
          >
            Quantity
          </td>
          <td
            role="columnheader"
          >
            Remove
          </td>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            mocked cart item 1
          </td>
          <td>
            <input
              class="update-input"
              value="4"
            />
          </td>
          <td>
            <button>
              <span
                class="fa fa-trash fa-2x"
              />
            </button>
          </td>
        </tr>
        <tr>
          <td>
            mocked cart item 2
          </td>
          <td>
            <input
              class="update-input"
              value="2"
            />
          </td>
          <td>
            <button>
              <span
                class="fa fa-trash fa-2x"
              />
            </button>
          </td>
        </tr>
        <tr>
          <td>
            mocked cart item 3
          </td>
          <td>
            <input
              class="update-input"
              value="7"
            />
          </td>
          <td>
            <button>
              <span
                class="fa fa-trash fa-2x"
              />
            </button>
          </td>
        </tr>
      </tbody>
    </table>
    <p
      class="actions"
    >
      <a
        href="/"
      >
        Continue shopping
      </a>
      <span>
        mock child fragment
        <button
          class="button-primary"
        >
          Place Order
        </button>
      </span>
    </p>
  </div>
</DocumentFragment>
`;
